### PR TITLE
Support concatenation of unpacked arrays

### PR DESCRIPTION
**Pull request by yTakatsukasa**
1800-2017 10.10.1 says

```systemverilog
A9 = {A3, 4, 5, A3, 6}; // legal, gives A9='{1,2,3,4,5,1,2,3,6}
```

where
```systemverilog
typedef int AI3[1:3];
AI3 A3;
int A9[1:9];
A3 = '{1, 2, 3};
```

This PR supports this. I push a test then will push code later as usual.

### CODE REVIEW

File: src/V3Width.cpp [hunk 1]
```diff
@@ -4129,7 +4130,7 @@ class WidthVisitor final : public VNVisitor {
                 if (defaultp) {
                     newpatp = defaultp->cloneTree(false);
                     patp = newpatp;
-                } else {
+                } else if (!VN_IS(arrayDtp, UnpackArrayDType) || allConstant) {
```
Comment (wsnyder) [inline src/V3Width.cpp:4133]: I'm not understanding... Perhaps add a comment?

**Review by wsnyder — COMMENTED** [review]
I think V3Slice is a bit of a mess, perhaps we should really be doing the slices more at runtime versus expanding the slice to many statements.  This will help side effect cases and probably others.  This isn't a reason to do it now, just commenting... Someday.

File: src/V3Slice.cpp [hunk 2]
```diff
@@ -56,14 +56,23 @@ class SliceVisitor final : public VNVisitor {
...
 
     // STATE
     AstNode* m_assignp = nullptr;  // Assignment we are under
     bool m_assignError = false;  // True if the current assign already has an error
     bool m_okInitArray = false;  // Allow InitArray children
 
     // METHODS
-    AstNodeExpr* cloneAndSel(AstNode* nodep, int elements, int offset) {
+
+    static bool areCompatible(const AstNodeDType* atypep, const AstNodeDType* btypep) {
```
[the remark this answers is not in the document]
Comment (yTakatsukasa) [reply src/V3Slice.cpp]: Thanks. I thought there must be a function like that, but I could not find it.

**Comment by yTakatsukasa** [comment]
Fixed range check. Now the test shows expected error messages.

**Pull request merged**